dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,161 @@
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith M.union ts1 ts2)
+
+data Tag = Tag
+    { tagDefinition :: Text
+    , tagName :: Text
+    }
+
+{-| Generate ETags for Dhall expressions
+
+    Arguments:
+
+    * Path to directory or filename that should be indexed
+    * List of suffixes for dhall files [".dhall"] by default
+    * Flag if generate should follow symlinks
+-}
+generate :: Input -> [Text] -> Bool -> IO Text
+generate StandardInput _ _ = showTags . tags "" <$> TIO.getContents
+generate (File p) sxs followSyms = do
+    isD <- doesDirectoryExist p
+    files <- dirToFiles followSyms (map T.unpack (if isD then sxs else [""])) p
+    showTags <$> foldM (\ts f -> do
+                                   t <- tags f <$> handle (\(SomeException _) -> return "")
+                                                          (TIO.readFile f)
+                                   return (ts <> t)) (Tags M.empty) files
+
+{-| Find tags in Text (second argument) and generates list of them
+    To make tags for filenames working in both, emacs and vi, add two initial tags.
+    First for `filename` for vi and second with `/filename` for emacs.
+    Other tags are working for both.
+-}
+tags :: FilePath -> Text -> Tags
+tags f t = Tags (M.singleton f
+                 (M.union initialMap (parse t)))
+    where initialMap = M.fromList [ (FP 1 0, Tag "" (T.pack . takeFileName $ f))
+                                  , (FP 1 1, Tag "" ("/" <> (T.pack . takeFileName) f))
+                                  ]
+
+parse :: Text -> M.Map FilePos Tag
+parse t = M.fromAscList . map (\(FP ln c, Tag _ term) ->
+              let (lsl, l) = mls M.! ln in
+              (FP ln (lsl + c), Tag (T.take c l) term)) . M.assocs $
+          parseExpr e (FP 0 0) M.empty
+    where e = case exprFromText "" t of
+                  Left _ -> None
+                  Right r -> r
+          mls = M.fromList . fst . foldl (\(ls, lsl) (n, l) ->
+                                             let lsl' = lsl + 1 + T.length l in
+                                             ((n, (lsl, l)):ls, lsl'))
+                                        ([], 0) . zip [1..] $ T.lines t
+
+parseExpr :: Expr Src a -> FilePos -> M.Map FilePos Tag -> M.Map FilePos Tag
+parseExpr (Let b e) lpos  mts = parseExpr e lpos (parseBinding b mts)
+parseExpr (Annot e1 e2) lpos mts = parseExpr e1 lpos (parseExpr e2 lpos mts)
+parseExpr (Record mr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $  mr
+parseExpr (RecordLit mr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $ mr
+parseExpr (Union mmr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $ mmr
+parseExpr (Note s e) _ mts = parseExpr e (srcToFilePos s) mts
+parseExpr _ _ mts = mts
+
+parseBinding :: Binding Src a -> M.Map FilePos Tag -> M.Map FilePos Tag
+parseBinding b = M.insert (maybeSrcToFilePos . bindingSrc0 $ b)
```

Comment:
Do you need the existing tags-map to produce the tags for a binding? If not, just return the `Map` with the tags for this bindings, and `mappend` outside of this function.

Reply:
Fixed.